File: boj/dp/java/Boj11659.java
import java.util.*;
import java.io.*;

public class Boj11659 {

    static int N, M;
    static int[] dp;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        dp = new int[N + 1];

        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= N; i++) {
            dp[i] = dp[i - 1] + Integer.parseInt(st.nextToken());
        }

        for (int k = 0; k < M; k++) {
            st = new StringTokenizer(br.readLine());
            int i = Integer.parseInt(st.nextToken());
            int j = Integer.parseInt(st.nextToken());
            sb.append(dp[j] - dp[i - 1]).append("\n");
        }

        System.out.print(sb);
    }
}

File: boj/dp/kotlin/Boj11659.kt
class Boj11659 {

    fun main() = with(System.`in`.bufferedReader()) {
        val (n, m) = readLine().split(" ").map { it.toInt() }
        val numbers = readLine().split(" ").map { it.toInt() }
        val prefixSum = Array(n) { numbers[it] }

        for (i in 1 until n) {
            prefixSum[i] += prefixSum[i - 1]
        }

        val sb = StringBuilder()
        repeat(m) {
            val (i, j) = readLine().split(" ").map { it.toInt() }
            sb.append(prefixSum[j - 1] - prefixSum[i - 1] + numbers[i - 1]).append("\n")
        }

        println(sb)
    }
}